package/gluon-mesh-vpn-wireguard/src/gluon-hex-to-b64.c

```diff
@@ -0,0 +1,102 @@
+// SPDX-FileCopyrightText: 2022 Jan-Niklas Burfeind <gluon@aiyionpri.me>
+// SPDX-License-Identifier: BSD-2-Clause
+// SPDX-FileContributor: read_hex() by Matthias Schiffer <mschiffer@universe-factory.net>
+
+#include <libubox/utils.h>
+#include <stdbool.h>
+#include <stdint.h>
+#include <stdio.h>
+#include <stdlib.h>
+#include <string.h>
+
+/** return how many samesized elements an array has */
+#define ARRAY_SIZE(arr) (sizeof(arr) / sizeof((arr)[0]))
+
+/**
+ * how many blocks should be encoded at once - can be configured
+ */
+#define BLOCK_AMOUNT 32
+
+/**
+ * smallest possible block size to encode in b64 without further contex
+ * is three bytes - do not change
+ */
+#define BLOCK_SIZE 3*BLOCK_AMOUNT
+
+/** print usage info and exit as failed */
+static void usage(void) {
+	fprintf(stderr, "Usage: gluon-hex-to-b64\n");
+	exit(1);
+}
+
+/**
+ * read a string of hexadecimal characters and return them as bytes
+ * return false in case any non-hexadecimal characters are provided
+ * return true on success
+ */
+static bool read_hex(uint8_t key[BLOCK_SIZE], const char *hexkey) {
+	if (strspn(hexkey, "0123456789abcdefABCDEF") != strlen(hexkey))
+		return false;
+
+	size_t i;
+	for (i = 0; i < BLOCK_SIZE; i++)
+		sscanf(&hexkey[2 * i], "%02hhx", &key[i]);
+
+	return true;
+}
+
+int main(int argc, char *argv[]) {
+	if (argc != 1)
+		usage();
+
+	unsigned char  hex_input[BLOCK_SIZE*2];
+	uint8_t as_bytes[BLOCK_SIZE];
+	int byte_count;
+	int b64_strlen = B64_ENCODE_LEN(BLOCK_SIZE);
+	int b64_return;
+	size_t ret;
+
+	char str[b64_strlen];
+
+	do {
+		ret = fread(hex_input, sizeof(*hex_input), ARRAY_SIZE(hex_input), stdin);
+
+		/* in case fread did not fill six characters */
+		if (ret != ARRAY_SIZE(hex_input)) {
+			/* drop newline by replacing it with a null character */
+			hex_input[strcspn(hex_input, "\n")] = 0;
+
+			/*
+			 * count length of resulting string and make sure it's even,
+			 * as bytes are represented using pairs of hex characters
+			 */
+			ret = strlen(hex_input);
+			if (1 == ret%2) {
+				fprintf(stderr, "Error: Incomplete hex representation of a byte.\n");
+				exit(EXIT_FAILURE);
+			}
+		}
+
+		byte_count = ret/2;
+		b64_strlen = B64_ENCODE_LEN(byte_count);
+
+		/* in case read_hex fails due to invalid characters */
+		if (!read_hex(as_bytes, hex_input)) {
+			fprintf(stderr, "Error: Invalid hexadecimal input.\n");
+			exit(EXIT_FAILURE);
+		}
+
+		b64_return = b64_encode(as_bytes, byte_count, str, b64_strlen);
+
+		if (b64_strlen - 1 != b64_return) {
+			fprintf(stderr, "Error: Encoding bytes as b64 failed.\n");
+			exit(EXIT_FAILURE);
+		}
+
+		printf("%s", str);
+	/* repeat until a non full block is read */
+	} while (ret == BLOCK_SIZE*2);
```

Comment:
```suggestion
	} while (ret == ARRAY_SIZE(hex_input));
```

For consistency with L65.

Reply:
resolved in 37d1ebaa898a3a06843d1d26b1ea424264691277